_posts/2022-11-25-async-sink-rate-limiting-strategy.md

```diff
@@ -0,0 +1,195 @@
+
+Side note: The default CongestionControlRateLimitingStrategy uses an Additive Increase / Multiplicative Decrease (AIMD) scaling strategy. This is also used in TCP congestion control to avoid overloading the destination by increasing write rate slowly but backing off quickly if throttled.
+
+```java
+public class CongestionControlRateLimitingStrategy implements RateLimitingStrategy {
+    // ...
+    @Override
+    public void registerCompletedRequest(ResultInfo resultInfo) {
+        // ...
+        if (resultInfo.getFailedMessages() > 0) {
+            maxInFlightMessages = scalingStrategy.scaleDown(maxInFlightMessages);
+        } else {
+            maxInFlightMessages = scalingStrategy.scaleUp(maxInFlightMessages);
+        }
+    }
+    
+    public boolean shouldBlock(RequestInfo requestInfo) {
+        return currentInFlightRequests >= maxInFlightRequests
+                || (currentInFlightMessages + requestInfo.getBatchSize() > maxInFlightMessages);
+    }
+    // ...
+}
+```
+
+### [Example 2] TokenBucketRateLimitingStrategy
+
+The CongestionControlRateLimitingStrategy is rather aggressive, and relies on a robust server-side rate limiting strategy. In the event we don’t have a robust server-side rate limiting strategy, we can implement a client-side rate limiting strategy.
+
+As an example, we can look at the [token bucket rate limiting strategy](https://en.wikipedia.org/wiki/Token_bucket). If the limits are set correctly, we will avoid overloading the destination altogether.
+
+In this strategy, we want to do the following:
+
+* Implement a TokenBucket that has a given initial number of tokens (e.g. 10). These tokens refill at a given rate (e.g. 1 token per second).
+* When preparing an async request, we check if the token bucket has sufficient tokens. If not, we postpone the request.
+
+Let’s look at an example implementation:
+
+```java
+public class TokenBucketRateLimitingStrategy implements RateLimitingStrategy {
+    
+    private final Bucket bucket;
+
+    public TokenBucketRateLimitingStrategy() {
+        Refill refill = Refill.intervally(1, Duration.ofSeconds(1));
+        Bandwidth limit = Bandwidth.classic(10, refill);
+        this.bucket = Bucket4j.builder()
+            .addLimit(limit)
+            .build();
+    }
+    
+    // ... (information methods not needed)
+    
+    @Override
+    public boolean shouldBlock(RequestInfo requestInfo) {
+        return bucket.tryConsume(requestInfo.getBatchSize());
+    }
+}
+```
+
+In the above example, we use the Token Bucket implementation in the [Bucket4j](https://github.com/bucket4j/bucket4j) library. We also map 1 message to 1 token. Since our token bucket has a size of 10 tokens and a refill rate of 1 token per second, we can be sure that we will not exceed a burst of 10 messages, and will also not exceed a constant rate of 1 message per second.
+
```

Comment:
I wonder if it makes sense to quickly summarize the characteristics of the strategy? It sound like this strategy really restricts the throughput to a certain rate.

Reply:
Sure, I can give some examples of how we could use this strategy to limit a specific byte/s or messages/s